src/v/serde/serde.h

```diff
@@ -91,6 +92,14 @@ concept has_serde_async_write = requires(T t, iobuf& out) {
     { t.serde_async_write(out) } -> seastar::Future;
 };
 
+template<typename T>
+concept is_absl_flat_hash_map = requires(T t) {
+    {
+        absl::flat_hash_map { t }
+    }
+    ->std::same_as<T>;
```

Comment:
this looks strange to me. is this invoking the copy constructor and then checking the result?

Reply:
Yeah it is weird. The general idea is to infer the template parameters, since we can't be sure that there aren't some that will be added at a later date that are also defaulted.

Reply:
Right, clang-format made it extra weird. I was leaning on using template parameters but then there are too many that are defaulted based on Key and mapped type and this seemed like the safest way.

Reply:
can you use `Args...` or list them explicitly (i think there are 4)? i'm not opposed to this but it still doesn't seem robust

Reply:
I was trying to do it directly with concepts, but we could probably just wait for this to merge: https://github.com/redpanda-data/redpanda/pull/5511
And then do this:
```c++
template<typename T>
concept is_absl_flat_hash_map
  = ::detail::is_specialization_of_v<T, abs::flat_hash_map>;
```

Reply:
> can you use `Args...` 

No, because you can't get at the name of template template parameter pack:
```c++
template<typename T, template<typename... Args> class C> # Args isn't usable below
concept is_specialization_of = std::is_same_v<T, C<Args...>>;
```

Reply:
> or list them explicitly (i think there are 4)? i'm not opposed to this but it still doesn't seem robust

This is fine for absl, but not for anything in std, such as `std::map`, since implementation are allowed to add an arbitrary number of default template parameters, so it might change over time, or between libc++ and libstdc++.

Reply:
>  i'm not opposed to this but it still doesn't seem robust

Can you expand a bit on this? I don't think they are evaluated like a normal expression (no overhead I mean). They are just unevaluated operands to infer type information (I'm pretty new to concepts too but this is my understanding of how they work).

Per [documentation](https://en.cppreference.com/w/cpp/experimental/constraints)
...
parameter-list - a comma-separated list of parameters like in a function declaration, except that default arguments are not allowed and the last parameter cannot be an ellipsis. These parameters have no storage, linkage or lifetime. 
....
A simple requirement is an arbitrary expression statement. The requirement is that the expression is valid (this is an expression constraint). Unlike with predicate constraints, evaluation does not take place,
...

Reply:
> > i'm not opposed to this but it still doesn't seem robust
> 
> Can you expand a bit on this?

It requires the type to be copyable, so it's not a good general purpose pattern. I should never had made the suggestion in the first place!

Reply:
Oh I thought it was about performance, my bad. Yes agree that this cannot be generalized. I can undo the change given we are building something generic in #5511.

Reply:
> Oh I thought it was about performance, my bad.

I wanted to do it in concepts directly because I thought it might be quicker, but we may as well move to the generic form that is a std proposal, which will make it easier to transition at a later date.

Reply:
what's the proposal here? old style sfinae stuff works...